cmd/buildkitd/main.go

```diff
@@ -196,7 +203,23 @@ func main() {
 				return err
 			}
 		}
-		opts := []grpc.ServerOption{unaryInterceptor(ctx), grpc.StreamInterceptor(otgrpc.OpenTracingStreamServerInterceptor(tracer))}
+		persistence := time.Duration(c.GlobalInt("persistence")) * time.Second
+		shutdownTimer := time.NewTimer(persistence)
+		if persistence == 0 {
+			shutdownTimer.Stop()
+		} else {
+			logrus.Warnf("the daemon will shutdown if no RPC activity in %v", persistence)
+		}
+		stopShutdownTimer := func() {
+			shutdownTimer.Stop()
+		}
+		resetShutdownTimer := func() {
+			if persistence != 0 {
+				logrus.Debugf("resetting shutdown timer (persistence=%v)", persistence)
+				shutdownTimer.Reset(persistence)
+			}
+		}
+		opts := []grpc.ServerOption{unaryInterceptor(ctx, stopShutdownTimer, resetShutdownTimer), streamInterceptor(stopShutdownTimer, resetShutdownTimer)}
```

Comment:
I'm not sure it is correct to detect idle in gRPC interceptors. Maybe we need to modify solver?